feat(Navbar): controlled subMenu

ShavrinAleksei opened the pull request:
## Описание изменений

## Чек-лист

- [ ] PR: направлен в правильную ветку
- [ ] PR: назначен исполнитель PR и указаны нужные лейблы
- [ ] PR: проверен diff, ничего лишнего в PR не попало
- [ ] PR: прилинкованы затронутые issue и связанные PR
- [ ] PR: есть описание изменений
- [ ] JS: нет варнингов и ошибок в консоли браузера
- [ ] Тесты: новый функционал и исправленные баги покрыты тестами
- [ ] Документация: отражены все изменения в API компонентов и описаны важные особенности реализации или использования
- [ ] Сторибук: для компонентов написаны или обновлены stories
- [ ] Верстка: используются [переменные](../src/components/Theme)
- [ ] Верстка: проверена с разным количеством контента

## Опционально

- [ ] Доработки: заведены задачи для дальнейшей работы, если что-то решено не править в текущем пулл-реквесте
- [ ] Коммиты: проименованы в соответствии с [правилами](https://consta-uikit.vercel.app/?path=/docs/common-develop-commits-style--page)

ShavrinAleksei committed:
chore(deps): update @consta/uikit version

ShavrinAleksei committed:
chore(package.json): add test script

ShavrinAleksei committed:
refactor(Navbar): remove unused NavbarDrawer

ShavrinAleksei committed:
feat(Navbar): add types for controlled subMenu

- Added `NavbarPropGetItemSubMenuOpen`
- Added `NavbarPropOnItemSubMenuToggle`
- Updated `NavbarProps` and `NavbarItemProps` to support controlled sub-menu behavior

ShavrinAleksei committed:
feat(Navbar): add passing of subMenu props

- Pass `getItemSubMenuOpen` and `onItemSubMenuToggle` to `NavbarItem`
- Update `withDefaultGetters`

ShavrinAleksei committed:
feat(Navbar): implement controlled subMenu state in NavbarItem

- Add controlled/uncontrolled logic via `getItemSubMenuOpen` and `onItemSubMenuToggle`
- Use external state when controlled, preserve local state otherwise
- Update click handlers and arrow toggle behavior
- Ensure backward compatibility

ShavrinAleksei committed:
chore(NavbarArrow): add tests for arrow component

ShavrinAleksei committed:
chore(NavbarRailItem): add tests for rail item component

ShavrinAleksei committed:
chore(NavbarRail): add tests for rail container component

ShavrinAleksei committed:
chore(NavbarItem): add tests for NavbarItem component

ShavrinAleksei committed:
chore(Navbar): add tests for Navbar component

ShavrinAleksei committed:
chore(jest.config.js): configure test setup

ShavrinAleksei committed:
docs(Navbar): add description and example for new subMenu props

ShavrinAleksei commented:
**Описание изменений**
Реализована возможность контролируемого управления состоянием подменю (subMenu) в компоненте Navbar. Теперь можно управлять открытием/закрытием подменю через внешний стейт, сохраняя обратную совместимость с текущей реализацией.

**Ключевые изменения**
1. Новая функциональность

- Добавлены пропсы getItemSubMenuOpen и onItemSubMenuToggle для контролируемого управления состоянием подменю
- Реализована гибридная логика: компонент работает в контролируемом режиме при передаче пропсов, и в неконтролируемом - при их отсутствии

2. Добавлены TypeScript типы

- NavbarPropGetItemSubMenuOpen
- NavbarPropOnItemSubMenuToggle

3. Обновлены NavbarProps и NavbarItemProps для поддержки новой функциональности

4. Добавлены комплексные тесты для всех компонентов Navbar

- Navbar (основной компонент)
- NavbarItem (элемент навигации)
- NavbarRail (рейл-контейнер)
- NavbarRailItem (элемент рейла)
- NavbarArrow (компонент стрелки)

5. Настроена конфигурация Jest для тестовой среды

6. Документация

- Добавлено описание новых пропсов и их использования
- Приведены примеры реализации контролируемого состояния подменю

7. Технические детали

- Обновлена версия @consta/uikit
- Удален неиспользуемый компонент NavbarDrawer
- Добавлен тестовый скрипт в package.json

**Закрыт issue:** Navbar: управление состоянием подменю #4063

gizeasy commented on src/components/Navbar/types.ts:
```diff
@@ -151,6 +161,8 @@ export type NavbarProps<
     sortGroup?: NavbarPropSortGroup<ITEM, GROUP>;
     getGroupAdditionalClassName?: NavbarPropGetGroupAdditionalClassName<GROUP>;
+    getItemSubMenuOpen?: NavbarPropGetItemSubMenuOpen<ITEM>;
+    onItemSubMenuToggle?: NavbarPropOnItemSubMenuToggle<ITEM>;
```
давай переименуем в `onSubMenuToggle`

gizeasy commented on src/components/Navbar/NavbarItem/NavbarItem.tsx:
```diff
@@ -51,24 +51,39 @@ const NavbarItemRender = (
 
   const [open, setOpen] = useFlag();
+  const isControlled = !!getItemSubMenuOpen && !!onItemSubMenuToggle;
+  const isOpen = isControlled ? getItemSubMenuOpen(item) ?? false : open;
```
кажется что немного перемудрил, 
что если не разделять 2 состояния внутреннее и внешнее и просто слушать всегда внутреннее. И синронить их серез useEffect

к примеру:
```tsx
// задаем начальное сотояние при первом рендере
const [open, setOpen] = useFlag(getItemSubMenuOpen?.(item) || false);

// при следующих рендерах дергаем пользовательское сотояние и если оно используется сетим его во внутреннее. 

const controlledOpen = getItemSubMenuOpen?.(item)

useEffect(() => {
  if(controlledOpen !== undefined){
    setOpen(controlledOpen)
  }
},[controlledOpen])
```

gizeasy commented on src/components/Navbar/NavbarItem/NavbarItem.tsx:
```diff
@@ -79,7 +94,12 @@ const NavbarItemRender = (
-            <NavbarArrow open={open} onClick={setOpen.toggle} />
+            <NavbarArrow
+              open={isOpen}
+              onClick={(e) => {
```
незачем создавать новую функцию

`onClick={handleToggle}`

gizeasy commented on src/components/Navbar/NavbarItem/NavbarItem.tsx:
```diff
@@ -51,24 +51,39 @@ const NavbarItemRender = (
+  const handleToggle = (e: React.MouseEvent) => {
+    if (subItems?.length) {
+      const newOpen = !isOpen;
+      if (isControlled) {
```
тут вне зависимости от isControlled возвращай в 
```tsx
onSubMenuToggle?.(!open)
setOpen.set(!open)
```

ShavrinAleksei committed:
refactor(Navbar): rename onItemSubMenuToggle to onSubMenuToggle

ShavrinAleksei committed:
refactor(Navbar): rename onItemSubMenuToggle to onSubMenuToggle

ShavrinAleksei committed:
refactor(NavbarItem): implement controlled subMenu with onSubMenuToggle

- Replaced onItemSubMenuToggle with onSubMenuToggle in handleToggle
- Added useEffect for syncing controlled open state
- Removed isControlled flag for simpler logic

ShavrinAleksei committed:
chore(Navbar): update tests for onSubMenuToggle rename and remove redundants

ShavrinAleksei committed:
docs(Navbar): update API reference for subMenu toggle changes

ShavrinAleksei committed:
docs(Navbar): update API reference for subMenu toggle types

gizeasy committed:
chore(deps): update

gizeasy commented on src/components/Navbar/types.ts:
```diff
@@ -96,6 +96,16 @@ export type NavbarPropSortGroup<ITEM, GROUP> = (
+
+export type NavbarPropOnSubMenuToggle<ITEM> = (
+  item: ITEM,
+  open: boolean,
```
переложи open в params, у нас устная договоренность в колбеках что первым аргументом идет item, а во втором аргументе объект со всем дополнительным

gizeasy commented on src/components/Navbar/NavbarItem/NavbarItem.tsx:
```diff
@@ -51,24 +51,41 @@ const NavbarItemRender = (
+  const [open, setOpen] = useFlag(getItemSubMenuOpen?.(item) || false);
+  const controlledOpen = getItemSubMenuOpen?.(item);
+
+  useEffect(() => {
```
хорошим тоном будет если описывать эффекты в самом низу перед рендером

ShavrinAleksei committed:
fix(Navbar): update onSubMenuToggle callback signature

ShavrinAleksei committed:
fix(NavbarItem): update onSubMenuToggle usage and reorder effects

ShavrinAleksei committed:
chore(Navbar): update tests for new onSubMenuToggle signature

ShavrinAleksei committed:
docs(Navbar): update docs for new onSubMenuToggle signature

gizeasy commented on src/components/Navbar/__stand__/Navbar.dev.stand.mdx:
````diff
@@ -115,6 +115,32 @@ export const NavbarOnClickExample = () => (
+
+Для управления состоянием подменю (открыто/закрыто) используйте связку getItemSubMenuOpen и onSubMenuToggle. Это позволяет контролировать открытие подменю через внешний стейт, включая задание начального состояния.
+
+```tsx
````
давай живой пример создадим который пощелкать можно будет.

менюха с 3мя пунктами 
- Пункт 1
- Пункт 2
- Пункт 3

и 3 кнопки 
- Открыть 1
- Открыть 2
- Открыть 3

если пункт открыт то кнопку переименуй в закрыть

gizeasy commented on src/components/Navbar/types.ts:
```diff
@@ -96,6 +96,18 @@ export type NavbarPropSortGroup<ITEM, GROUP> = (
 
+export type NavbarPropGetItemSubMenuOpen<ITEM> = (
+  item: ITEM,
+) => boolean | undefined;
```
`undefined` думаю стоит убрать так как далее по логике у нас 

```
const controlledOpen = getItemSubMenuOpen?.(item);

useEffect(() => {
    if (controlledOpen !== undefined) {
      setOpen.set(controlledOpen);
    }
  }, [controlledOpen]);

```

где если приходит undefined то мы считаем что контролировать пункт не надо и не будем синкать внутреннее состояние. А пользователь может подумать что если не true то будет закрываться, и отдаст undefined. Пусть будет строго boolean

ShavrinAleksei committed:
fix(Navbar): change getItemSubMenuOpen return type from boolean | undefined to strict boolean

ShavrinAleksei committed:
docs(Navbar): update getItemSubMenuOpen type documentation

ShavrinAleksei committed:
feat(Navbar): add controlled submenu example

ShavrinAleksei committed:
docs(Navbar): add interactive controlled submenu example

ShavrinAleksei committed:
feat(Navbar): add CSS styles for controlled submenu example

ShavrinAleksei committed:
refactor(Navbar): update controlled example to use CSS classes

ShavrinAleksei committed:
docs(Navbar): add interactive tabs for controlled submenu example

gizeasy committed:
docs(Navbar): edit menu
